felix/dataplane/linux/bpf_ep_mgr.go

```diff
@@ -4201,10 +4211,246 @@ func (m *bpfEndpointManager) getIfaceTypeFromLink(link netlink.Link) IfaceType {
 }
 
 func (m *bpfEndpointManager) getBondSlaves(masterIfName string) set.Set[string] {
-	if val, ok := m.hostIfaceToSlaveDevices[masterIfName]; ok {
+	slaveDevices := set.New[string]()
+	link, err := m.dp.getIfaceLink(masterIfName)
+	if err != nil {
+		log.Errorf("Error getting link for interface %s, err = %v", masterIfName, err)
+		return slaveDevices
+	}
+	hostIf := m.findHostIfaceByIndex(link.Attrs().Index)
+	if hostIf == nil {
+		log.Errorf("error finding interface %s", masterIfName)
+		return slaveDevices
+	}
+
+	if len(hostIf.children) > 0 {
+		leaves := []string{}
+		getLeafNodes(hostIf, &leaves)
+		slaveDevices.AddAll(leaves)
+	}
+	return slaveDevices
+}
+
+// addHostIface adds host interface to hostIfaces tree.
+func (m *bpfEndpointManager) addHostIface(link netlink.Link) {
+	attrs := link.Attrs()
+	intf := &bpfHostIface{name: attrs.Name,
+		index:       attrs.Index,
+		masterIndex: attrs.MasterIndex,
+		parentIndex: attrs.ParentIndex,
+		children:    make(map[int]*bpfHostIface)}
+
+	if attrs.MasterIndex == 0 && attrs.ParentIndex == 0 {
+		eintf := m.findHostIfaceByIndex(attrs.Index)
+		if eintf != nil {
+			eintf.name = intf.name
+			eintf.index = intf.index
+			eintf.masterIndex = intf.masterIndex
+			eintf.parentIndex = intf.parentIndex
+		}
+		if val, exists := m.hostIfaces[attrs.Index]; !exists {
+			m.hostIfaces[attrs.Index] = intf
+		} else {
+			val.name = intf.name
+			val.index = intf.index
+			val.masterIndex = intf.masterIndex
+			val.parentIndex = intf.parentIndex
+		}
+	} else if attrs.MasterIndex != 0 {
+		m.deleteHostIface(intf.name)
+		// If the master interface is in the tree, add it as a slave.
+		if val, exists := m.hostIfaces[attrs.MasterIndex]; exists {
+			intf.parentIface = val
+			val.children[attrs.Index] = intf
+			m.hostIfaces[attrs.MasterIndex] = val
+		} else {
+			// Need to search slave devices.
+			masterIf := m.findHostIfaceByIndex(attrs.MasterIndex)
+			if masterIf != nil {
+				intf.parentIface = masterIf
+				masterIf.children[attrs.Index] = intf
+			} else {
+				masterIface := &bpfHostIface{index: attrs.MasterIndex, children: make(map[int]*bpfHostIface)}
+				intf.parentIface = masterIface
+				masterIface.children[attrs.Index] = intf
+				m.hostIfaces[attrs.MasterIndex] = masterIface
+			}
+		}
+		delete(m.hostIfaces, attrs.Index)
```

Comment:
why are you deleting it if you just added it and it could be the new root?

Reply:
When adding an interface with masterIndex !=0 (which means the intf is a slave intf), we need to delete the interface from the tree before adding it because the interface might have moved from standalone to a bond. Hence the deletion.